File: python/isbn-verifier/isbn_verifier.py
def is_valid(isbn):
    ''' 
    Checks if a given ISBN-10 is valid.
    Input (string): ISBN to test
    Output (bool): True if valid, otherwise False
    '''
       
    # Check valid length and remove hyphen
    isbn = isbn.replace("-", "")
    if len(isbn) != 10:
        return False

    multiplicand = 10
    checksum = 0
    for char in isbn:

        if char.isdigit():
            checksum += int(char) * multiplicand

        elif char == "X" and multiplicand == 1:
            # Only the last "digit" can be X: We can reuse the multiplicand var
            # to check for this
            checksum += 10

        else:
            # Not a valid character
            return False
        multiplicand -= 1
    
    return checksum % 11 == 0

File: java/isbn-verifier/src/main/java/IsbnVerifier.java
class IsbnVerifier {

    boolean isValid(String stringToVerify) {
        // Remove everything but digits.
        String digits = stringToVerify.replaceAll("[^0-9X ]", "");

        if (digits.length() != 10)
            return false;

        int checksum = 0;
        int multiplicand = 10;
        for (int i = 0; i < 9; i++) {
            char d = digits.charAt(i);

            if (!Character.isDigit(d))
                return false;

            checksum += Character.getNumericValue(d) * multiplicand;
            multiplicand--;
        }

        // Last digit is a special case.
        if (digits.charAt(9) == 'X')
            checksum += 10;
        else
            checksum += Character.getNumericValue(digits.charAt(9));

        return checksum % 11 == 0;
    }
}
